src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -625,102 +610,73 @@ namespace xclhwemhal2 {
     if (xclemulation::config::getInstance()->isSharedFmodel() && !mVersalPlatform) {
       setenv("SDX_USE_SHARED_MEMORY","true",true);
     }
-    // iterate cores
-    count = 0;
-    for (auto& xml_core : xml_project.get_child("project.platform.device"))
-    {
-      if (xml_core.first != "core")
-        continue;
-      if (++count > 1)
-      {
-        //Give error and return from here
-      }
-    }
 
-    std::vector<std::string> kernels;
-
-    // iterate kernels
-    for (auto& xml_kernel : xml_project.get_child("project.platform.device.core"))
+    auto xclbin_object = new xrt::xclbin(args.m_top);
+    auto kernels = xclbin_object->get_kernels();
+    std::map<uint64_t, KernelArg> kernelArgInfo;
+    std::string instance_name = "";
+    uint64_t base_address = 0;
+    for (auto kernel : kernels)
     {
-      if (xml_kernel.first != "kernel")
-        continue;
-
-      std::string kernelName = xml_kernel.second.get<std::string>("<xmlattr>.name");
-      kernels.push_back(kernelName);
-      int address_range = 0;
-      std::string instanceName;
-
-      if (mLogStream.is_open())
-         mLogStream << __func__ << " Filling kernel " << kernelName << " info from xclbin.xml" << std::endl;
-
-      for (auto& xml_kernel_info : xml_kernel.second)
+      // get properties of each kernel object
+      auto props = xrt_core::xclbin_int::get_properties(kernel);
+      //get CU's of each kernel object
+      auto cus = kernel.get_cus();
+      //iterate over CU's to get arguments
+      for (auto cu :cus)
       {
-        std::map<uint64_t, KernelArg> kernelArgInfo;
-        if (xml_kernel_info.first == "arg")
+        auto args = cu.get_args();
+        base_address = cu.get_base_address();
+        mCuBaseAddress = base_address & 0xFFFFFFFF00000000;
+        //BAD Worharound for vck5000 need to remove once SIM_QDMA supports PCIE bar
+        if(xclemulation::config::getInstance()->getCuBaseAddrForce()!=-1)
         {
-          std::string name = xml_kernel_info.second.get<std::string>("<xmlattr>.name");
-          std::string id = xml_kernel_info.second.get<std::string>("<xmlattr>.id");
-          std::string port = xml_kernel_info.second.get<std::string>("<xmlattr>.port");
-          uint64_t offset = convert(xml_kernel_info.second.get<std::string>("<xmlattr>.offset"));
-          uint64_t size = convert(xml_kernel_info.second.get<std::string>("<xmlattr>.size"));
-          KernelArg kArg;
-          kArg.name = kernelName + ":" + name;
-          kArg.size = size;
-          kernelArgInfo[offset] = kArg;
-
-          if (mLogStream.is_open())
-            mLogStream << __func__ << " Filling kernel Args name: " << name << " id: " << id << " port: " << port << " info from xclbin.xml" << std::endl;
+          mCuBaseAddress = xclemulation::config::getInstance()->getCuBaseAddrForce();
+        } 
+        else if(mVersalPlatform)
```

Comment:
Conditional code for platform type doesn't look right, it is definitely not scalable; this should be considered for redesign.